### ISSUE TITLE
surefire.forkNumber undefined variable

### ISSUE THREAD

**Issue by r6squeegee** [eclipse-m2e/m2e-core#1884]
surefire.forkNumber generates an error when junit test is run from with in eclipse.

Basically in a pom.xml surefire plugin section, you can configure to fork. This works on the command line using mvn directly, this also worked in prior versions of eclipse. I'm unsure when it broke but I think it was with in the last month or so.

Reverting back to a much earlier version of eclipse (the only one I had was 2024-03) shows it worked back then.

The reason I didn't notice it the day it happened was because you must right click the project -> maven -> update project, before it breaks (or after you install a new eclipse), if your system was working with an older version it would continue working until you had to do an update-project.

attached as a simple project showing this.

![dialog](https://github.com/user-attachments/assets/6b8ac8cc-5177-4e99-afe9-97a30b86caca)

[temp.zip](https://github.com/user-attachments/files/17605193/temp.zip)

To un-break it, you have to downgrade to an older version of eclipse and delete the working directories .metadata directory or it will continue to appear broken even on older versions.

We also tested jdk 17,21,23 and it happens on all versions. It happened on multiple fedora 40 machines. It happened with both the eclipse-installer with bundled jdk and binary untar-gz versions with separately downloaded openjdks.

the download was tagged eclipse-jee-2024-09-R-linux-gtk-x86_64.tar.gz

**Comment by merks** [comment]
Have you tried my suggestion  to define a corresponding variable in Window -> Preferences -> Run/Debug -> String Substitutions. That **definitely** will solve the problem.

**Comment by r6squeegee** [comment]
the problem isn't working around it, the problem is a regression in functionality.

I believe I can also work around it by adding -DDEV to the jvm run parameters.

**Comment by merks** [comment]
Have you looked at the launch configuration itself?  I expect it will contain the `${surefire.forkNumber}` in the VM or Progres args section.  It may well be that adding this is a regression or maybe it's an enhancement.   I just wanted to be sure your not blocked until there is an answer and/or a fix...

**Comment by r6squeegee** [comment]
yes, in the run configuration -> arguments -> vm arguments it populates that with 
```
-DforkQualifier=_${surefire.forkNumber}
```

and yes, if I manually remove that, it will run again.

The easier work around is for me to just set the forking to 1 and hard code the fork count instead of using the forkNumber right now. The reason being - that vm argument is auto generated /added by eclipse every time you run a junit test for the first time. 

So effectively every time some one runs a junit test, instead of just right clicking -> run as junit, they would have to right-click -> run junit, watch it fail, right-click -> run configuration -> remove argument -> run.

Thanks for the concern about this blocking me. I'm not blocked, I've just hard coded it to 1 thread and the number 1, and it  works for all our developers with out any additional effort for any of them. (the performance difference for forking v.s. non-forking was not significant for us anyways.)

I just wanted to lodge this issue as it seems like no one else had run into it yet (and it took us quite a while to narrow it down to this, we though we broke it with something we did). It also means that anyone else using any surefire variables (or possibly any other plugins with any other variables) will also no longer work for the same reason. So while my problem was trivial to work around, it may be more significant for other people.

**Comment by treilhes** [comment]
Hello @r6squeegee,

M2e create/update the test configuration on "Maven update project" or when running the test for the first time.
surefire.forkNumber is a dynamic parameter unknown to m2e during configuration creation.

I think that providing a default value using properties and overriding the value at runtime would do the trick.

```
<properties>
	<surefire.forkNumber>1</surefire.forkNumber>
</properties>
```

Using cli, what happen when you don't provide the argument -DforkQualifier ?

**Comment by merks** [comment]
In the ide you can define a string substitution in the preferences as a workaround.

**Comment by r6squeegee** [comment]
on the CLI if I just run ./mvnw clean verify, I don't have to provide -DforkQualifier, I only have to do that with in the eclipse configuration. Everything works fine on the command line, this is a normal feature of surefire and it works well.

So I just wanted to re-iterate the concern. I can work around this easily by just not forking my tests, so this doesn't actually cause me many issues as I have a work around. I just submitted this so everyone is aware of it and it's documented as it may affect other people more severely as this may be a larger new problem with how systemPropertyVariables are passed to the IDE.

The concern is that a proper/normal pom.xml with normal maven-surefire-plugin setup using forkCount, use to automatically be detected / configured and worked properly with in eclipse. That is no longer the case. (or was no longer the case, I haven't re-tested this with the current build.) As for manual work-arounds with in eclipse, there are a few as outlined prior and they work, but if there's a many people working on the project, it's much easier to have things just-work as they should, rather than have every developer add a work-around into their ide, then have to remember to constantly re-apply the work around any time they resetup or reconfigure their machine etc. Given this hassle, it was much easier to stop using forkCounts and have a few extra seconds delay in getting our results, than have to remember and document a work around for everyones ide configration.

**Comment by treilhes** [comment]
Hello @jtnord,
What is exactly your issue ? What i'm guessing is that defining a default value using:

```
<properties>
	<surefire.forkNumber>1</surefire.forkNumber>
</properties>
```

breaks your jenkins pipeline, am i right ?
